Source/ASDisplayNode.mm

```diff
@@ -2118,6 +2119,26 @@ - (ASDisplayNode *)supernode
   return _supernode;
 }
 
+// Repeatedly: try lock self, try lock root node, unlock self,
+// return the lock on root node. If self is root once we've locked it,
+// we'll return the lock on self.
+- (ASDN::UniqueLock)acquireRootLock {
+  for (;; std::this_thread::yield()) {
+    ASDN::UniqueLock selfLock(__instanceLock__, std::try_to_lock);
+    if (!selfLock.owns_lock()) {
+      continue;
+    }
+    if (_rootNode == self) {
+      return selfLock;
+    }
+    ASDN::UniqueLock rootLock(_rootNode->__instanceLock__, std::try_to_lock);
+    if (!rootLock.owns_lock()) {
+      continue;
+    }
+    return rootLock;
+  }
+}
+
 - (void)_setSupernode:(ASDisplayNode *)newSupernode
```

Comment:
YogaTree is different from normal node tree.  Meaning when we create a yoga tree, the rootnode will always be self.  So even during yoga layout in calculateLayouFromYogaRoot the _rootNode may not be updated timely.